### PR TITLE
Add support for configuring WiFi devices

### PR DESCRIPTION
**Pull request by teclator**
## Problem

In #260 we added basic support for configuring the network, the idea is to extend it supporting also wireless devices.

- https://trello.com/c/SFpwJaU7/3139-5-d-installer-add-support-for-configuring-wifi-devices

## Solution

We have added support for configuring WiFi networks at least allowing the user to connect to a network without security or using **WPA-PSK** authentication as well as configuring a **Hidden network**.

## Testing

- Added some unit tests **(but it is still not well covered)**.
- Tested manually

## Screenshots

<details>
<summary>Click to show/hide some screenshots</summary>

---

  | Network summary with Wi-Fi capabilities | Available networks | Connecting to a network |
  |-|-|-|
  | Installation summary holding a network section with Wi-Fi capabilities | Available Wi-Fi networks for connecting to | Connecting to a Wi-Fi network |

  | Form for setting a connection | Form for setting a hidden connection | Detecting errors BEFORE trying to connect |
  |-|-|-|
  | Form for setting a new Wi-Fi connection | Form for setting a hidden connection | Detecting errors BEFORE trying to connect |

### CODE REVIEW

File: web/src/WirelessSelector.jsx [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    { value: "wpa-psk", label: "WPA & WPA2 Personal" },
+    { value: "wpa-eap", label: "WPA & WPA2 Enterprise" }
+  ];
+
+  const selectorOptions = security_options.map(security => (
+    <FormSelectOption key={security.value} value={security.value} label={security.label} />
+  ));
+
+  const connectNetwork = async () => {
+    await client.network.connectTo(ssid, { security, password });
```
Comment (dgdavid) [inline web/src/WirelessSelector.jsx:67]: @teclator,

At this point, we have to send the hidden option when needed, right?

File: web/src/WifiSelector.jsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    });
+  });
+
+  const isStateChanging = (network) => {
+    const state = network.connection?.state;
+    return state === ConnectionState.ACTIVATING || state === ConnectionState.DEACTIVATING;
+  };
+
+  const renderFilteredNetworks = (networks) => {
+    return networks.map(n => {
```
Comment (imobachgs) [inline web/src/WifiSelector.jsx:171]: Please, use a meaningful name. For a small function, is perfectly fine, but this one has around 60 lines :-)

File: web/src/WifiSelector.jsx [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+              Connect to hidden network
+            </Button>
+          </Center> }
+      </>
+    );
+  };
+
+  return (
+    <Popup isOpen={isOpen} height="large" title="Connect to a Wi-Fi network">
+      { renderFilteredNetworks(networksFromValues(networks)) }
```
Comment (imobachgs) [inline web/src/WifiSelector.jsx:266]: Why not use a proper component? OK, I know, components are functions, but I would expect to have something like `Networks` and `HiddenNetworkForm`. Do you need to share all that state between them?

In any case, about the `renderFilteredNetworks`, does that function do any filtering? If that's not the case, I do not think you need to mention that in the name. It isn't very clear.

File: web/src/WifiSelector.jsx [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    const loadNetworks = async () => {
+      const knownSsids = [];
+
+      return client.network.accessPoints()
+        .sort((a, b) => b.strength - a.strength)
+        .reduce((networks, ap) => {
+          // Do not include networks without SSID
+          if (!ap.ssid || ap.ssid === "") return networks;
+          // Do not include "duplicates"
+          if (knownSsids.includes(ap.ssid)) return networks;
```
Comment (imobachgs) [inline web/src/WifiSelector.jsx:93]: What about filtering networks you cannot configure? For instance, we cannot set up a WEP network, right? Alternatively, we could show them, but we should not allow trying to connect.

**Review by imobachgs — CHANGES_REQUESTED** [review]

File: web/src/WifiConnectionForm.jsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  FormGroup,
+  FormSelect,
+  FormSelectOption,
+  TextInput
+} from "@patternfly/react-core";
+import { useInstallerClient } from "./context/installer";
+
+const security_options = [
+  { value: "", label: "None" },
+  { value: "wpa-psk", label: "WPA & WPA2 Personal" }
```
[the remark this answers is not in the document]
Comment (teclator) [reply web/src/WifiConnectionForm.jsx:37]: In this case we missed a class **enum** for adding things like the label for example.

File: web/src/WifiConnectionForm.jsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+];
+
+const selectorOptions = security_options.map(security => (
+  <FormSelectOption key={security.value} value={security.value} label={security.label} />
+));
+
+const securityFrom = (supported) => {
+  if (supported.includes("WPA2"))
+    return "wpa-psk";
```
[the remark this answers is not in the document]
Comment (teclator) [reply web/src/WifiConnectionForm.jsx:46]: Similar to above and it also shows that is very related to the adapter... so maybe it should be declared there.

File: web/src/TargetIpsPopup.jsx [hunk 7]
```diff
@@ -30,64 +30,45 @@ import { formatIp } from "./client/network";
...
 
   const open = () => setIsOpen(true);
   const close = () => setIsOpen(false);
 
   return (
     <>
-      <Button variant="link" onClick={open} isDisabled={ips.length === 1}>
-        {firstIp} {hostname && <Text component="small">({hostname})</Text>}
+      <Button variant="link" onClick={open} isDisabled={addresses.length === 1}>
+        {formatIp(firstIp)} {hostname && <Text component="small">({hostname})</Text>}
```
[the remark this answers is not in the document]
Comment (teclator) [reply web/src/TargetIpsPopup.jsx:65]: Better to move now it than later ;)

File: web/src/WifiSelector.jsx (hunk 3 above)
Comment (dgdavid) [reply web/src/WifiSelector.jsx]: Imo, you're right.

Somehow, we forget to extract this to small components after so many changes to get the Wi-Fi selector working. I have [tried to improve the situation a bit](https://github.com/yast/d-installer/pull/292/commits/beb0e4a7ffcf41c5d41e1c4facb9c546e2028472), but I fear a better improvement/re-organization must wait until next iterations in the network area.

What is more, I'd like to create a custom _SelectionList_ component (Wi-Fi and Product selector are quite similar, though) and reorganize not only that part, but the whole src/ directory. 

Hope, however, it looks good/acceptable enough for now.

File: web/src/HiddenNetworkForm.jsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import WifiConnectionForm from "./WifiConnectionForm";
+
+function HiddenNetworkForm({ network, expanded, onClick, onCancel, onSubmitCallback }) {
+  return (
+    <>
+      <Card className={[
+        "selection-list-item",
+        expanded && "selection-list-focused-item",
+        !expanded && "collapsed"
+      ].join(" ")}
```
Comment (dgdavid) [inline web/src/HiddenNetworkForm.jsx:42]: NOTE: better to use a util function to improve this and avoid _"false"_ CSS classes. 

See https://github.com/jangouts/jangouts/blob/2081d8f344b016a884b58f651a46aa69ece5070a/src/utils/common.js#L8-L18

```js
/**
 * Helper to simplify building the className prop conditionally
 *
 * @param {String} CSS classes to join
 * @returns {String} final CSS classes properly joined, exluding "false"s
 */
function classNames(...classes) {
  return classes.filter((item) => !!item).join(' ');
}
```

File: web/src/WifiConnectionForm.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/WifiConnectionForm.jsx:37]: Yes, that's true. My point is that using a "wpa-psk" looks fragile. But let's address it later.

File: web/src/WifiConnectionForm.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/WifiConnectionForm.jsx:46]: Sure it should not be declared here :-) but as said before, we can live with this.

File: web/src/NetworkListItem.jsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    case ConnectionState.DEACTIVATING:
+      return 'Disconnecting';
+    case ConnectionState.DEACTIVATED:
+      return 'Disconnected';
+    default:
+      return "";
+  }
+};
+
+function NetworkListItem ({ network, isChecked, isFocused, showAsChecked, showSpinner, onSelect, onCancel }) {
```
Comment (imobachgs) [inline web/src/NetworkListItem.jsx:60]: It would be nice to document the arguments.

File: web/src/NetworksList.jsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+function NetworksList({ networks, activeNetwork, selectedNetwork, onSelectionCallback, onCancelSelectionCallback }) {
+  const isStateChanging = (network) => {
+    const state = network.connection?.state;
+    return state === ConnectionState.ACTIVATING || state === ConnectionState.DEACTIVATING;
+  };
+
+  return networks.map(n => {
+    const isChecked = n.ssid === selectedNetwork?.ssid;
+    const showAsChecked = !selectedNetwork && n.ssid === activeNetwork?.ssid;
```
Comment (imobachgs) [inline web/src/NetworksList.jsx:35]: np: Would not be better something like `isActive`? I meant, the component will decide whether to show it. It sounds more "semantic" (perhaps `isSelected` and `isActive`). Just a suggestion, feel free to ignore it.

File: web/src/NetworksList.jsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    const showSpinner = (isChecked && n.settings && !n.connection) || isStateChanging(n);
```
Comment (imobachgs) [inline web/src/NetworksList.jsx:36]: The same as above: the `NetworkListItem` should decide which is the right representation (a spinner, a different color, etc.).

**Review by imobachgs — APPROVED** [review]
It looks much better now. Just a few suggestions, but feel free to ignore them.

File: web/src/NetworksList.jsx (hunk 10 above)
Comment (dgdavid) [reply web/src/NetworksList.jsx:35]: Completely agree. Not an _np_ at at all.

Believe it or not, I thought the same after reading the code, a couple of hours after writing it.

Definitely, being in a rush is not a good thing ;)

Thanks a lot for taking your time for carefully reviewing that PR.

File: web/src/NetworkListItem.jsx (hunk 9 above)
Comment (dgdavid) [reply web/src/NetworkListItem.jsx:60]: Ok, let's do it

File: web/src/HiddenNetworkForm.jsx (hunk 8 above)
Comment (dgdavid) [reply web/src/HiddenNetworkForm.jsx]: Done, see https://github.com/yast/d-installer/pull/292/commits/4491ee2c0651c589f7a1e6add535549e7b825be1

File: web/src/ConnectionsDataList.test.jsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  name: "WiFi 1",
+  type: ConnectionTypes.WIFI,
+  addresses: [{ address: "192.168.69.200", prefix: 24 }]
+};
+
+let conns = [];
+
+describe("ConnectionsDataList", () => {
+  describe("when no connections are given", () => {
+    it("renders nothing", async () => {
```
Comment (imobachgs) [inline web/src/ConnectionsDataList.test.jsx:47]: It is unnecessary to make this test async (IMHO, it is wrong). Actually, `ConnectionDataList` is a pretty simple component: no async, no `useInstallerClient`...

Please, have a look at #313 which should simplify the test. Of course, do not hesitate to ask if anything remains unclear.

File: web/src/ConnectionsDataList.test.jsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      const { container } = installerRender(<ConnectionsDataList conns={conns} />, { usingLayout: false });
```
Comment (imobachgs) [inline web/src/ConnectionsDataList.test.jsx:48]: You can simply pass an empty array here. So you do not need to reassign `conns` in the test below.

File: web/src/WifiHiddenNetworkForm.test.jsx [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+import WifiHiddenNetworkForm from "./WifiHiddenNetworkForm";
+
+const beforeDisplayingFn = jest.fn();
+
+jest.mock("./WifiConnectionForm", () => () => "WifiConnectionForm mock");
+
+describe("WifiHiddenNetworkForm", () => {
+  describe("when it is visible", () => {
+    it("renders the WifiConnectionForm", async () => {
```
Comment (imobachgs) [inline web/src/WifiHiddenNetworkForm.test.jsx:35]: No need for async in `WifiHiddenNetworkForm` tests (except for the last one which tests user interaction). Please, check #313 for an example.

File: web/src/WifiHiddenNetworkForm.test.jsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      installerRender(<WifiHiddenNetworkForm visible />);
```
Comment (imobachgs) [inline web/src/WifiHiddenNetworkForm.test.jsx:36]: Please, use `plainRender` as `useInstallerClient` is not needed.

File: web/src/WifiNetworkListItem.test.jsx [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+};
+
+const fakeSettings = {
+  wireless:  {
+    password: "notSecret"
+  }
+};
+
+describe("NetworkListItem", () => {
+  it("renders an input radio for selecting the network", async () => {
```
Comment (imobachgs) [inline web/src/WifiNetworkListItem.test.jsx:46]: In general, there is no need for async (except where user interaction is tested).

File: web/src/WifiNetworkListItem.test.jsx [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import React from "react";
+
+import { screen } from "@testing-library/react";
+import { installerRender } from "./test-utils";
```
Comment (imobachgs) [inline web/src/WifiNetworkListItem.test.jsx:25]: I think you can use `plainRender`.

File: web/src/WifiNetworksList.test.jsx [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+const otherNetwork = {
+  ssid: "My Neighbour Network",
+  security: ["WPA2"],
+  strengh: 35
+};
+
+const networksMock = [myNetwork, otherNetwork];
+
+describe("WifiNetworksList", () => {
```
Comment (imobachgs) [inline web/src/WifiNetworksList.test.jsx:46]: No need for async.

File: web/src/WifiNetworksList.test.jsx [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import React from "react";
+
+import { screen, waitFor } from "@testing-library/react";
+import { installerRender } from "./test-utils";
```
Comment (imobachgs) [inline web/src/WifiNetworksList.test.jsx:25]: You can use `plainRender`.

**Review by imobachgs — CHANGES_REQUESTED** [review]
I have reviewed the tests more closely and found out that they use async too much. Additionally, if you do need to access the `useInstallerClient` hook, you can use `plainRender` instead.

File: web/src/ConnectionsDataList.test.jsx (hunk 13 above)
Comment (teclator) [reply web/src/ConnectionsDataList.test.jsx:48]: Yep, originally it was an empty array, then I expected to write more tests and contexts and moved it to a variable, but finally it is not needed at all.

**Review by imobachgs — APPROVED** [review]

**Pull request merged**